src/events/attributes.rs

```diff
@@ -572,6 +474,427 @@ impl std::error::Error for AttrError {}
+    /// [`AttrError::ExpectedEq`] will be returned instead
+    Empty(T),
+}
+
+impl<T> Attr<T> {
+    /// Maps an `Attr<T>` to `Attr<U>` by applying a function to a contained key and value.
+    #[inline]
+    pub fn map<U, F>(self, mut f: F) -> Attr<U>
+    where
+        F: FnMut(T) -> U,
+    {
+        match self {
+            Attr::DoubleQ(key, value) => Attr::DoubleQ(f(key), f(value)),
+            Attr::SingleQ(key, value) => Attr::SingleQ(f(key), f(value)),
+            Attr::Empty(key) => Attr::Empty(f(key)),
+            Attr::Unquoted(key, value) => Attr::Unquoted(f(key), f(value)),
+        }
+    }
+}
+
+impl<'a> Attr<&'a [u8]> {
+    /// Returns the key value
+    #[inline]
+    pub fn key(&self) -> &'a [u8] {
+        match self {
+            Attr::DoubleQ(key, _) => key,
+            Attr::SingleQ(key, _) => key,
+            Attr::Empty(key) => key,
+            Attr::Unquoted(key, _) => key,
+        }
+    }
+    /// Returns the attribute value. For [`Self::Empty`] variant an empty slice
+    /// is returned according to the [HTML specification].
+    ///
+    /// [HTML specification]: https://www.w3.org/TR/2012/WD-html-markup-20120329/syntax.html#syntax-attr-empty
+    #[inline]
+    pub fn value(&self) -> &'a [u8] {
+        match self {
+            Attr::DoubleQ(_, value) => value,
+            Attr::SingleQ(_, value) => value,
+            Attr::Empty(_) => &[],
+            Attr::Unquoted(_, value) => value,
+        }
+    }
+}
+
+impl<T: AsRef<[u8]>> Debug for Attr<T> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Attr::DoubleQ(key, value) => f
+                .debug_tuple("Attr::DoubleQ")
+                .field(&Bytes(key.as_ref()))
+                .field(&Bytes(value.as_ref()))
+                .finish(),
+            Attr::SingleQ(key, value) => f
+                .debug_tuple("Attr::SingleQ")
+                .field(&Bytes(key.as_ref()))
+                .field(&Bytes(value.as_ref()))
+                .finish(),
+            Attr::Empty(key) => f
+                .debug_tuple("Attr::Empty")
+                // Comment to prevent formatting and keep style consistent
+                .field(&Bytes(key.as_ref()))
+                .finish(),
+            Attr::Unquoted(key, value) => f
+                .debug_tuple("Attr::Unquoted")
+                .field(&Bytes(key.as_ref()))
+                .field(&Bytes(value.as_ref()))
+                .finish(),
+        }
+    }
+}
+
+/// Unpacks attribute key and value into tuple of this two elements.
+/// `None` value element is returned only for [`Attr::Empty`] variant.
+impl<T> From<Attr<T>> for (T, Option<T>) {
+    fn from(attr: Attr<T>) -> Self {
+        match attr {
+            Attr::DoubleQ(key, value) => (key, Some(value)),
+            Attr::SingleQ(key, value) => (key, Some(value)),
+            Attr::Empty(key) => (key, None),
+            Attr::Unquoted(key, value) => (key, Some(value)),
+        }
+    }
+}
```

Comment:
Basically all this code would be deleted up to my previous comment.  `key()` and `value()` methods would stay but it would just be a simple accessor.